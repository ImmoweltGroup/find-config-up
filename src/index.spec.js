// @flow

const findConfigUp = require('./index.js');

const mockFs = {
  '/foo/raw-config/nested/folder/some.json': '{}',
  '/foo/raw-config/nested/package.json': '{}',
  '/foo/raw-config/.foorc': '{"someRawRcProperty": true}',

  '/foo/package-json/nested/folder/package.json': '{}',
  '/foo/package-json/nested/package.json': '{}',
  '/foo/package-json/package.json': '{"foo": {"somePackageJsonProperty": true}}'
};

describe('findConfigUp()', () => {
  let readFileAsync;
  let findUp;

  beforeEach(() => {
    readFileAsync = jest
      .spyOn(findConfigUp._utils, 'readFileAsync')
      .mockImplementation(jest.fn(path => mockFs[path]));
    findUp = jest
      .spyOn(findConfigUp._utils, 'findUp')
      .mockImplementation(jest.fn());
  });

  afterEach(() => {
    readFileAsync.mockRestore();
    findUp.mockRestore();
  });

  it('should be a function', () => {
    expect(typeof findConfigUp).toBe('function');
  });

  it('should return the contents of the provided "rawConfigFileName" filename if one was found up the filesystem tree', async () => {
    findUp.mockReturnValueOnce('/foo/raw-config/.foorc');

    const cfg = await findConfigUp({
      rawConfigFileName: '.foorc',
      packageJsonProperty: 'foo',
      cwd: '/foo/raw-config/nested/folder'
    });

    expect(cfg).toMatchSnapshot();
  });

  it('should return the contents of the "package.json" with the given "packageJsonProperty" property if one was found up the filesystem tree', async () => {
    findUp
      .mockReturnValueOnce(null)
      .mockReturnValueOnce('/foo/package-json/nested/folder/package.json')
      .mockReturnValueOnce('/foo/package-json/nested/package.json')
      .mockReturnValue('/foo/package-json/package.json');

    const cfg = await findConfigUp({
      rawConfigFileName: '.foorc',
      packageJsonProperty: 'foo',
      cwd: '/foo/package-json/nested/folder'
    });

    expect(cfg).toMatchSnapshot();
  });

  it('should return the defaults if no file was found', async () => {
    findUp.mockReturnValue(null);

    const cfg = await findConfigUp({
      rawConfigFileName: '.foorc',
      packageJsonProperty: 'foo',
      cwd: '/foo/package-json/nested/folder',
      defaults: {someDefaultsProperty: true}
    });

    expect(cfg).toMatchSnapshot();
  });

  it('should not crash if no cwd was given', () => {
    findUp.mockReturnValue(null);

    expect(() =>
      findConfigUp({
        rawConfigFileName: '.foorc',
        packageJsonProperty: 'foo'
      })
    ).not.toThrow();
  });
});
