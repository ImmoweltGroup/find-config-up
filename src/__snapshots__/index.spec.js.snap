// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`findConfigUp() should return the contents of the "package.json" with the given "packageJsonProperty" property if one was found up the filesystem tree 1`] = `
Object {
  "somePackageJsonProperty": true,
}
`;

exports[`findConfigUp() should return the contents of the provided "rawConfigFileName" filename if one was found up the filesystem tree 1`] = `
Object {
  "someRawRcProperty": true,
}
`;

exports[`findConfigUp() should return the defaults if no file was found 1`] = `
Object {
  "someDefaultsProperty": true,
}
`;
